File: Tests/Source/Data Model/ZMConversationTests+messages.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
// 


#import "ZMConversationTests.h"
#import "ZMClientMessage.h"


@implementation ZMConversationTests (Messages)

- (void)testThatWeCanInsertATextMessage;
{
    [self.syncMOC performGroupedBlockAndWait:^{
        // given
        ZMUser *selfUser = [ZMUser selfUserInContext:self.syncMOC];
        ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.syncMOC];
        conversation.remoteIdentifier = NSUUID.createUUID;
        
        // when
        NSString *messageText = @"foo";
        id<ZMConversationMessage> message = [conversation appendMessageWithText:messageText];
        
        // then
        XCTAssertEqualObjects(message.messageText, messageText);
        XCTAssertEqual(message.conversation, conversation);
        XCTAssertTrue([conversation.messages containsObject:message]);
        XCTAssertEqualObjects(selfUser, message.sender);
    }];
}

- (void)testThatItUpdatesTheLastModificationDateWhenInsertingMessagesIntoAnEmptyConversation;
{
    // given
    ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.uiMOC];
    conversation.lastModifiedDate = [[NSDate date] dateByAddingTimeInterval:-90000];
    ZMMessage *msg = [conversation appendMessageWithText:@"Foo"];
    
    // then
    XCTAssertNotNil(msg.serverTimestamp);
    XCTAssertEqualObjects(conversation.lastModifiedDate, msg.serverTimestamp);
}

- (void)testThatItUpdatesTheLastModificationDateWhenInsertingMessages;
{
    // given
    ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.uiMOC];
    ZMMessage *msg1 = [conversation appendMessageWithText:@"Foo"];
    msg1.serverTimestamp = [[NSDate date] dateByAddingTimeInterval:-90000];
    conversation.lastModifiedDate = msg1.serverTimestamp;
    
    // when
    ZMMessage *msg2 = [conversation appendMessageWithImageData:[self verySmallJPEGData]];
    
    // then
    XCTAssertNotNil(msg2.serverTimestamp);
    XCTAssertEqualObjects(conversation.lastModifiedDate, msg2.serverTimestamp);
}

- (void)testThatItIsSafeToPassInAMutableStringWhenCreatingATextMessage
{
    // given
    ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.uiMOC];
    conversation.remoteIdentifier = NSUUID.createUUID;
    
    // when
    NSString *originalText = @"foo";
    NSMutableString *messageText = [NSMutableString stringWithString:originalText];
    id<ZMConversationMessage> message = [conversation appendMessageWithText:messageText];
    
    // then
    [messageText appendString:@"1234"];
    XCTAssertEqualObjects(message.messageText, originalText);
}

- (void)testThatInsertATextMessageWithNilTextDoesNotCreateANewMessage
{
    // given
    ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.uiMOC];
    conversation.remoteIdentifier = NSUUID.createUUID;
    NSSet *start = [self.uiMOC.insertedObjects copy];
    
    // when
    __block ZMTextMessage *message;
    [self performIgnoringZMLogError:^{
        message = [conversation appendMessageWithText:nil];
    }];
    
    // then
    XCTAssertNil(message);
    XCTAssertEqualObjects(start, self.uiMOC.insertedObjects);
}

- (void)testThatWeCanInsertAnImageMessageFromAFileURL;
{
    // given
    ZMUser *selfUser = [ZMUser selfUserInContext:self.uiMOC];
    NSURL *imageFileURL = [self fileURLForResource:@"1900x1500" extension:@"jpg"];
    ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.uiMOC];
    conversation.remoteIdentifier = NSUUID.createUUID;
    
    // when
    ZMImageMessage *message = (ZMImageMessage *)[conversation appendMessageWithImageAtURL:imageFileURL];
    
    // then
    XCTAssertNotNil(message);
    XCTAssertNotNil(message.nonce);
    XCTAssertTrue(CGSizeEqualToSize(message.originalSize, CGSizeMake(1900, 1500)));
    XCTAssertEqual(message.conversation, conversation);
    XCTAssertTrue([conversation.messages containsObject:message]);
    XCTAssertNotNil(message.nonce);
    NSData *expectedData = [NSData dataWithContentsOfURL:imageFileURL];
    XCTAssertNotNil(expectedData);
    AssertEqualData(message.originalImageData, expectedData);
    XCTAssertEqualObjects(selfUser, message.sender);
}

- (void)testThatNoMessageIsInsertedWhenTheImageFileURLIsPointingToSomethingThatIsNotAnImage;
{
    // given
    NSURL *imageFileURL = [self fileURLForResource:@"1900x1500" extension:@"jpg"];
    ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.uiMOC];
    conversation.remoteIdentifier = NSUUID.createUUID;
    
    // when
    ZMImageMessage *message = (ZMImageMessage *)[conversation appendMessageWithImageAtURL:imageFileURL];
    
    // then
    XCTAssertNotNil(message);
    XCTAssertNotNil(message.nonce);
    AssertEqualSizes(message.originalSize, CGSizeMake(1900, 1500));
    XCTAssertEqual(message.conversation, conversation);
    XCTAssertTrue([conversation.messages containsObject:message]);
    XCTAssertNotNil(message.nonce);
    NSData *expectedData = [NSData dataWithContentsOfURL:imageFileURL];
    XCTAssertNotNil(expectedData);
    AssertEqualData(message.originalImageData, expectedData);
}

- (void)testThatNoMessageIsInsertedWhenTheImageFileURLIsNotAFileURL
{
    // given
    NSURL *imageURL = [NSURL URLWithString:@"http://www.placehold.it/350x150"];
    ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.uiMOC];
    conversation.remoteIdentifier = NSUUID.createUUID;
    NSSet *start = [self.uiMOC.insertedObjects copy];
    
    // when
    __block ZMImageMessage *message;
    [self performIgnoringZMLogError:^{
        message = (ZMImageMessage *)[conversation appendMessageWithImageAtURL:imageURL];
    }];
    
    // then
    XCTAssertNil(message);
    XCTAssertEqualObjects(start, self.uiMOC.insertedObjects);
}

- (void)testThatNoMessageIsInsertedWhenTheImageFileURLIsNotPointingToAFile
{
    // given
    NSURL *textFileURL = [self fileURLForResource:@"Lorem Ipsum" extension:@"txt"];
    XCTAssertNotNil(textFileURL);
    ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.uiMOC];
    conversation.remoteIdentifier = NSUUID.createUUID;
    NSSet *start = [self.uiMOC.insertedObjects copy];
    
    // when
    __block ZMImageMessage *message;
    [self performIgnoringZMLogError:^{
        message = (ZMImageMessage *)[conversation appendMessageWithImageAtURL:textFileURL];
    }];
    
    // then
    XCTAssertNil(message);
    XCTAssertEqualObjects(start, self.uiMOC.insertedObjects);
}

- (void)testThatWeCanInsertAnImageMessageFromImageData;
{
    // given
    NSData *imageData = [self dataForResource:@"1900x1500" extension:@"jpg"];
    XCTAssertNotNil(imageData);
    ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.uiMOC];
    conversation.remoteIdentifier = NSUUID.createUUID;
    
    // when
    ZMImageMessage *message = (ZMImageMessage *)[conversation appendMessageWithImageData:imageData];
    
    // then
    XCTAssertNotNil(message);
    XCTAssertNotNil(message.nonce);
    XCTAssertTrue(CGSizeEqualToSize(message.originalSize, CGSizeMake(1900, 1500)));
    XCTAssertEqual(message.conversation, conversation);
    XCTAssertTrue([conversation.messages containsObject:message]);
    XCTAssertNotNil(message.nonce);
    AssertEqualData(message.originalImageData, imageData);
}

- (void)testThatItIsSafeToPassInMutableDataWhenCreatingAnImageMessage
{
    // given
    NSData *originalImageData = [self dataForResource:@"1900x1500" extension:@"jpg"];
    NSMutableData *imageData = [originalImageData mutableCopy];
    XCTAssertNotNil(imageData);
    ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.uiMOC];
    conversation.remoteIdentifier = NSUUID.createUUID;
    
    // when
    ZMImageMessage *message = (ZMImageMessage *)[conversation appendMessageWithImageData:imageData];
    
    // then
    [imageData appendBytes:((const char []) {1, 2}) length:2];
    AssertEqualData(message.originalImageData, originalImageData);
}

- (void)testThatNoMessageIsInsertedWhenTheImageDataIsNotAnImage;
{
    // given
    NSData *textData = [self dataForResource:@"Lorem Ipsum" extension:@"txt"];
    XCTAssertNotNil(textData);
    ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.uiMOC];
    conversation.remoteIdentifier = NSUUID.createUUID;
    NSSet *start = [self.uiMOC.insertedObjects copy];
    
    // when
    __block ZMImageMessage *message;
    [self performIgnoringZMLogError:^{
        message = (ZMImageMessage *)[conversation appendMessageWithImageData:textData];
    }];
    
    // then
    XCTAssertNil(message);
    XCTAssertEqualObjects(start, self.uiMOC.insertedObjects);
}

- (void)testThatLastReadUpdatesInSelfConversationDontExpire
{
    [self.syncMOC performGroupedBlockAndWait:^{
        // given
        ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.syncMOC];
        conversation.remoteIdentifier = NSUUID.createUUID;
        conversation.lastReadServerTimeStamp = [NSDate date];
        
        // when
        ZMClientMessage *message = [ZMConversation appendSelfConversationWithLastReadOfConversation:conversation];
        
        // then
        XCTAssertNotNil(message);
        XCTAssertNil(message.expirationDate);
    }];
}

- (void)testThatLastClearedUpdatesInSelfConversationDontExpire
{
    [self.syncMOC performGroupedBlockAndWait:^{
        // given
        ZMConversation *conversation = [ZMConversation insertNewObjectInManagedObjectContext:self.syncMOC];
        conversation.remoteIdentifier = NSUUID.createUUID;
        conversation.clearedTimeStamp = [NSDate date];
        
        // when
        ZMClientMessage *message = [ZMConversation appendSelfConversationWithClearedOfConversation:conversation];
        
        // then
        XCTAssertNotNil(message);
        XCTAssertNil(message.expirationDate);
    }];
}

@end // messages

File: wire-ios-data-model/Tests/Source/Model/Conversation/ZMConversationTests+Messages.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import WireImages
import XCTest
@testable import WireDataModel

final class ZMConversationMessagesTests: ZMConversationTestsBase {
    func testThatWeCanInsertATextMessage() {
        syncMOC.performGroupedAndWait {
            // given
            let selfUser = ZMUser.selfUser(in: self.syncMOC)
            let conversation = ZMConversation.insertNewObject(in: self.syncMOC)
            conversation.remoteIdentifier = UUID()

            // when
            let messageText = "foo"
            let message = try! conversation.appendText(content: messageText) as! ZMMessage

            // then
            XCTAssertEqual(message.textMessageData?.messageText, messageText)
            XCTAssertEqual(message.conversation, conversation)
            XCTAssertEqual(conversation.lastMessage as! ZMMessage, message)
            XCTAssertEqual(selfUser, message.sender)
        }
    }

    func testThatItUpdatesTheLastModificationDateWhenInsertingMessagesIntoAnEmptyConversation() {
        // given
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.lastModifiedDate = Date(timeIntervalSinceNow: -90000)

        // when
        guard let msg = try? conversation.appendText(content: "Foo") as? ZMMessage else {
            XCTFail()
            return
        }

        // then
        XCTAssertNotNil(msg.serverTimestamp)
        XCTAssertEqual(conversation.lastModifiedDate, msg.serverTimestamp)
    }

    func testThatItUpdatesTheLastModificationDateWhenInsertingMessages() {
        // given
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        guard let msg1 = try? conversation.appendText(content: "Foo") as? ZMMessage else {
            XCTFail()
            return
        }
        msg1.serverTimestamp = Date(timeIntervalSinceNow: -90000)
        conversation.lastModifiedDate = msg1.serverTimestamp

        // when
        guard let msg2 = try? conversation.appendImage(from: verySmallJPEGData()) as? ZMAssetClientMessage else {
            XCTFail()
            return
        }

        // then
        XCTAssertNotNil(msg2.serverTimestamp)
        XCTAssertEqual(conversation.lastModifiedDate, msg2.serverTimestamp)
    }

    func testThatItDoesNotUpdateTheLastModifiedDateForRenameAndLeaveSystemMessages() {
        let types = [
            ZMSystemMessageType.teamMemberLeave,
            ZMSystemMessageType.conversationNameChanged,
            ZMSystemMessageType.messageTimerUpdate,
        ]

        for type in types {
            // given
            let conversation = ZMConversation.insertNewObject(in: uiMOC)
            let lastModified = Date(timeIntervalSince1970: 10)
            conversation.lastModifiedDate = lastModified

            let systemMessage = ZMSystemMessage(nonce: UUID(), managedObjectContext: uiMOC)
            systemMessage.systemMessageType = type
            systemMessage.serverTimestamp = lastModified.addingTimeInterval(100)

            // when
            conversation.append(systemMessage)

            // then
            XCTAssertEqual(conversation.lastModifiedDate, lastModified)
        }
    }

    func testThatItIsSafeToPassInAMutableStringWhenCreatingATextMessage() {
        // given
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        // when
        let originalText = "foo"
        var messageText = originalText
        let message = try! conversation.appendText(content: messageText)

        // then
        messageText.append("1234")
        XCTAssertEqual(message.textMessageData?.messageText, originalText)
    }

    func testThatWeCanInsertAnImageMessageFromAFileURL() {
        // given
        let selfUser = ZMUser.selfUser(in: uiMOC)
        let imageFileURL = fileURL(forResource: "1900x1500", extension: "jpg")!
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        // when
        let message = try! conversation.appendImage(at: imageFileURL) as! ZMAssetClientMessage

        // then
        XCTAssertNotNil(message)
        XCTAssertNotNil(message.nonce)
        XCTAssertTrue(message.imageMessageData!.originalSize.equalTo(CGSize(width: 1900, height: 1500)))
        XCTAssertEqual(message.conversation, conversation)
        XCTAssertEqual(conversation.lastMessage as! ZMMessage, message)
        XCTAssertNotNil(message.nonce)

        let expectedData = try! (try! Data(contentsOf: imageFileURL)).wr_removingImageMetadata()
        XCTAssertNotNil(expectedData)
        XCTAssertEqual(message.imageMessageData?.imageData, expectedData)
        XCTAssertEqual(selfUser, message.sender)
    }

    func testThatNoMessageIsInsertedWhenTheImageFileURLIsPointingToSomethingThatIsNotAnImage() {
        // given
        let imageFileURL = fileURL(forResource: "1900x1500", extension: "jpg")!
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        // when
        let message = try! conversation.appendImage(at: imageFileURL) as! ZMAssetClientMessage

        // then
        XCTAssertNotNil(message)
        XCTAssertNotNil(message.nonce)
        XCTAssertTrue(message.imageMessageData!.originalSize.equalTo(CGSize(width: 1900, height: 1500)))
        XCTAssertEqual(message.conversation, conversation)
        XCTAssertEqual(conversation.lastMessage as! ZMMessage, message)
        XCTAssertNotNil(message.nonce)

        let expectedData = try! (try! Data(contentsOf: imageFileURL)).wr_removingImageMetadata()
        XCTAssertNotNil(expectedData)
        XCTAssertEqual(message.imageMessageData?.imageData, expectedData)
    }

    func testThatNoMessageIsInsertedWhenTheImageFileURLIsNotAFileURL() {
        // given
        let imageURL = URL(string: "http://www.placehold.it/350x150")!
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()
        let start = uiMOC.insertedObjects

        // when
        var message: Any?
        performIgnoringZMLogError {
            message = try? conversation.appendImage(at: imageURL)
        }

        // then
        XCTAssertNil(message)
        XCTAssertEqual(start, uiMOC.insertedObjects)
    }

    func testThatNoMessageIsInsertedWhenTheImageFileURLIsNotPointingToAFile() {
        // given
        let textFileURL = fileURL(forResource: "Lorem Ipsum", extension: "txt")!
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()
        let start = uiMOC.insertedObjects

        // when
        var message: Any?
        performIgnoringZMLogError {
            message = try? conversation.appendImage(at: textFileURL)
        }

        // then
        XCTAssertNil(message)
        XCTAssertEqual(start, uiMOC.insertedObjects)
    }

    // swiftlint:disable:next todo_requires_jira_link
    // TODO: check why fail on Xcode 11
    func disable_testThatWeCanInsertAnImageMessageFromImageData() {
        // given
        let imageData = try! data(forResource: "1900x1500", extension: "jpg").wr_removingImageMetadata()
        XCTAssertNotNil(imageData)
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        // when
        guard let message = try? conversation.appendImage(from: imageData) as? ZMAssetClientMessage else {
            XCTFail()
            return
        }

        // then
        XCTAssertNotNil(message)
        XCTAssertNotNil(message.nonce)
        XCTAssertTrue(message.imageMessageData!.originalSize.equalTo(CGSize(width: 1900, height: 1500)))
        XCTAssertEqual(message.conversation, conversation)
        XCTAssertEqual(conversation.lastMessage as! ZMMessage, message)
        XCTAssertNotNil(message.nonce)
        XCTAssertEqual(message.imageMessageData?.imageData?.count, imageData.count)
    }

    // swiftlint:disable:next todo_requires_jira_link
    // TODO: check why fail on Xcode 11
    func disable_testThatItIsSafeToPassInMutableDataWhenCreatingAnImageMessage() {
        // given
        let originalImageData = try! data(forResource: "1900x1500", extension: "jpg").wr_removingImageMetadata()
        var imageData = originalImageData
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        // when
        guard let message = try? conversation.appendImage(from: imageData) as? ZMAssetClientMessage else {
            XCTFail()
            return
        }

        // then
        imageData.append(contentsOf: [1, 2])
        XCTAssertEqual(message.imageMessageData?.imageData?.count, originalImageData.count)
    }

    func testThatNoMessageIsInsertedWhenTheImageDataIsNotAnImage() {
        // given
        let textData = data(forResource: "Lorem Ipsum", extension: "txt")!
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()
        let start = uiMOC.insertedObjects

        // when
        var message: ZMConversationMessage?
        performIgnoringZMLogError {
            message = try? conversation.appendImage(from: textData)
        }

        // then
        XCTAssertNil(message)
        XCTAssertEqual(start, uiMOC.insertedObjects)
    }

    func testThatLastReadUpdatesInSelfConversationDontExpire() {
        syncMOC.performGroupedAndWait {
            // given
            let conversation = ZMConversation.insertNewObject(in: self.syncMOC)
            conversation.remoteIdentifier = UUID()
            conversation.lastReadServerTimeStamp = Date()

            // when
            guard let message = try? ZMConversation.updateSelfConversation(withLastReadOf: conversation) else {
                XCTFail()
                return
            }

            // then
            XCTAssertNil(message.expirationDate)
        }
    }

    func testThatWeCanInsertAFileMessage() {
        // given
        let documents = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first!
        let fileURL = URL(fileURLWithPath: documents).appendingPathComponent("secret_file.txt")
        let data = Data.randomEncryptionKey()
        let size = data.count
        try! data.write(to: fileURL)
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        // when
        let fileMetaData = ZMFileMetadata(fileURL: fileURL)
        let fileMessage = try! conversation.appendFile(with: fileMetaData) as! ZMAssetClientMessage

        // then
        XCTAssertEqual(conversation.lastMessage as! ZMMessage, fileMessage)

        XCTAssertNotNil(fileMessage)
        XCTAssertNotNil(fileMessage.nonce)
        XCTAssertNotNil(fileMessage.fileMessageData)
        XCTAssertNotNil(fileMessage.underlyingMessage)
        XCTAssertNil(fileMessage.assetId)
        XCTAssertFalse(fileMessage.delivered)
        XCTAssertTrue(fileMessage.hasDownloadedFile)
        XCTAssertEqual(fileMessage.size, UInt64(size))
        XCTAssertEqual(fileMessage.progress, 0)
        XCTAssertEqual(fileMessage.filename, "secret_file.txt")
        XCTAssertEqual(fileMessage.mimeType, "text/plain")
        XCTAssertFalse(fileMessage.fileMessageData!.isVideo)
        XCTAssertFalse(fileMessage.fileMessageData!.isAudio)
    }

    func testThatWeCanNotInsertAFileMessage_WhenFileSharingIsDisabled() {
        // given
        let documents = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first!
        let fileURL = URL(fileURLWithPath: documents).appendingPathComponent("secret_file.txt")
        let data = Data.randomEncryptionKey()
        try! data.write(to: fileURL)
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        // when
        let fileSharingFeature = Feature.fetch(name: .fileSharing, context: uiMOC)
        fileSharingFeature?.status = .disabled
        let fileMetaData = ZMFileMetadata(fileURL: fileURL)

        do {
            _ = try conversation.appendFile(with: fileMetaData) as! ZMAssetClientMessage
        } catch let error as NSError {
            // then
            XCTAssertEqual(error as! ZMConversation.AppendMessageError, .fileSharingIsRestricted)
        }
    }

    func testThatWeCanInsertATextMessageWithImageQuote() {
        // given
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()
        let imageMessage = try? conversation.appendImage(from: verySmallJPEGData())

        // when
        let textMessage = try? conversation.appendText(content: "Hello World", replyingTo: imageMessage)

        // then
        XCTAssertNotNil(textMessage?.textMessageData?.quoteMessage)
        XCTAssertEqual(textMessage?.textMessageData?.quoteMessage?.nonce, imageMessage?.nonce)
    }

    func testThatWeCanInsertAPassFileMessage() {
        // given
        let filename = "ticket.pkpass"
        let documents = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first!
        let fileURL = URL(fileURLWithPath: documents).appendingPathComponent(filename)
        let data = Data.randomEncryptionKey()
        let size = data.count
        try! data.write(to: fileURL)
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        // when
        let fileMetaData = ZMFileMetadata(fileURL: fileURL)
        let fileMessage = try! conversation.appendFile(with: fileMetaData) as! ZMAssetClientMessage

        // then
        XCTAssertEqual(conversation.lastMessage as! ZMMessage, fileMessage)

        XCTAssertNotNil(fileMessage)
        XCTAssertNotNil(fileMessage.nonce)
        XCTAssertNotNil(fileMessage.fileMessageData)
        XCTAssertNotNil(fileMessage.underlyingMessage)
        XCTAssertNil(fileMessage.assetId)
        XCTAssertFalse(fileMessage.delivered)
        XCTAssertTrue(fileMessage.hasDownloadedFile)
        XCTAssertEqual(fileMessage.size, UInt64(size))
        XCTAssertEqual(fileMessage.progress, 0)
        XCTAssertEqual(fileMessage.filename, filename)
        XCTAssertEqual(fileMessage.mimeType, "application/vnd.apple.pkpass")
        XCTAssertFalse(fileMessage.fileMessageData!.isVideo)
        XCTAssertFalse(fileMessage.fileMessageData!.isAudio)
        XCTAssert(fileMessage.fileMessageData!.isPass)
    }

    func locationData() -> LocationData {
        let latitude = Float(48.53775)
        let longitude = Float(9.041169)
        let zoomLevel = Int32(16)
        let name = "天津市 နေပြည်တော် Test"
        return LocationData(
            latitude: latitude,
            longitude: longitude,
            name: name,
            zoomLevel: zoomLevel
        )
    }

    func testThatWeCanInsertALocationMessage() {
        // given
        let latitude = Float(48.53775)
        let longitude = Float(9.041169)
        let zoomLevel = Int32(16)
        let name = "天津市 နေပြည်တော် Test"
        let locationData = locationData()

        // when
        syncMOC.performGroupedAndWait {
            let conversation = ZMConversation.insertNewObject(in: self.syncMOC)
            conversation.remoteIdentifier = UUID()

            guard let message = try? conversation.appendLocation(with: locationData) as? ZMMessage else {
                XCTFail()
                return
            }

            XCTAssertEqual(conversation.lastMessage as! ZMMessage, message)

            guard let locationMessageData = message.locationMessageData else {
                XCTFail()
                return
            }
            XCTAssertEqual(locationMessageData.longitude, longitude)
            XCTAssertEqual(locationMessageData.latitude, latitude)
            XCTAssertEqual(locationMessageData.zoomLevel, zoomLevel)
            XCTAssertEqual(locationMessageData.name, name)
        }
    }

    func testThatLocationMessageHasNoImage() throws {
        // given
        let locationData = locationData()

        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.setMessageDestructionTimeoutValue(.fiveMinutes, for: .selfUser)
        conversation.remoteIdentifier = UUID()
        // when
        let message = try conversation.appendLocation(with: locationData) as! ZMClientMessage

        // then
        XCTAssertNil(message.underlyingMessage?.imageAssetData)
        XCTAssertNotNil(message.underlyingMessage?.locationData)
        XCTAssertTrue(message.shouldExpire)
    }

    func testThatWeCanInsertAVideoMessage() {
        // given
        let fileName = "video.mp4"
        let documents = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first!
        let fileURL = URL(fileURLWithPath: documents).appendingPathComponent(fileName)
        let videoData = Data.secureRandomData(length: 500)
        let thumbnailData = Data.secureRandomData(length: 250)
        let duration = 12333
        let dimensions = CGSize(width: 1900, height: 800)
        try! videoData.write(to: fileURL)

        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        // when
        let videoMetadata = ZMVideoMetadata(
            fileURL: fileURL,
            duration: TimeInterval(duration),
            dimensions: dimensions,
            thumbnail: thumbnailData
        )

        guard let fileMessage = try? conversation.appendFile(with: videoMetadata) as? ZMAssetClientMessage else {
            XCTFail()
            return
        }

        // then
        XCTAssertEqual(conversation.lastMessage as! ZMMessage, fileMessage)

        XCTAssertNotNil(fileMessage)
        XCTAssertNotNil(fileMessage.nonce)
        XCTAssertNotNil(fileMessage.fileMessageData)
        XCTAssertNotNil(fileMessage.underlyingMessage)
        XCTAssertNil(fileMessage.assetId)
        XCTAssertFalse(fileMessage.delivered)
        XCTAssertTrue(fileMessage.hasDownloadedFile)
        XCTAssertEqual(fileMessage.size, UInt64(videoData.count))
        XCTAssertEqual(fileMessage.progress, 0)
        XCTAssertEqual(fileMessage.filename, fileName)
        XCTAssertEqual(fileMessage.mimeType, "video/mp4")
        guard let fileMessageData = fileMessage.fileMessageData else {
            XCTFail()
            return
        }
        XCTAssertTrue(fileMessageData.isVideo)
        XCTAssertFalse(fileMessageData.isAudio)
        XCTAssertEqual(fileMessageData.durationMilliseconds, UInt64(duration * 1000))
        XCTAssertEqual(fileMessageData.videoDimensions.height, dimensions.height)
        XCTAssertEqual(fileMessageData.videoDimensions.width, dimensions.width)
    }

    func testThatWeCanInsertAnAudioMessage() {
        // given
        let fileName = "audio.m4a"
        let documents = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first!
        let fileURL = URL(fileURLWithPath: documents).appendingPathComponent(fileName)
        let videoData = Data.secureRandomData(length: 500)
        let thumbnailData = Data.secureRandomData(length: 250)
        let duration = 12333
        try! videoData.write(to: fileURL)

        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        // when
        let audioMetadata = ZMAudioMetadata(
            fileURL: fileURL,
            duration: TimeInterval(duration),
            normalizedLoudness: [],
            thumbnail: thumbnailData
        )

        let fileMessage = try! conversation.appendFile(with: audioMetadata) as! ZMAssetClientMessage

        // then
        XCTAssertEqual(conversation.lastMessage as! ZMMessage, fileMessage)

        XCTAssertNotNil(fileMessage)
        XCTAssertNotNil(fileMessage.nonce)
        XCTAssertNotNil(fileMessage.fileMessageData)
        XCTAssertNotNil(fileMessage.underlyingMessage)
        XCTAssertNil(fileMessage.assetId)
        XCTAssertFalse(fileMessage.delivered)
        XCTAssertTrue(fileMessage.hasDownloadedFile)
        XCTAssertEqual(fileMessage.size, UInt64(videoData.count))
        XCTAssertEqual(fileMessage.progress, 0)
        XCTAssertEqual(fileMessage.filename, fileName)
        XCTAssertEqual(fileMessage.mimeType, "audio/x-m4a")
        guard let fileMessageData = fileMessage.fileMessageData else {
            XCTFail()
            return
        }
        XCTAssertFalse(fileMessageData.isVideo)
        XCTAssertTrue(fileMessageData.isAudio)
    }

    func testThatItDoesNotFetchMessageWhenMissing() {
        // GIVEN
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        // WHEN
        let lastMessage = conversation.lastMessageSent(by: selfUser)

        // THEN
        XCTAssertEqual(lastMessage, nil)
    }

    func testThatItFetchesMessageForUser() {
        // GIVEN
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        let message = try! conversation.appendText(content: "Test Message") as! ZMMessage

        // WHEN
        let lastMessage = conversation.lastMessageSent(by: selfUser)

        // THEN
        XCTAssertEqual(lastMessage, message)
    }

    func testThatItFetchesLastMessageForUser() {
        // GIVEN
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        _ = try! conversation.appendText(content: "Test Message") as! ZMMessage
        let message2 = try! conversation.appendText(content: "Test Message 2") as! ZMMessage

        // WHEN
        let lastMessage = conversation.lastMessageSent(by: selfUser)

        // THEN
        XCTAssertEqual(lastMessage, message2)
    }

    func testThatItIgnoreMessagesFromOtherUsers() {
        // GIVEN
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()

        let message1 = try! conversation.appendText(content: "Test Message") as! ZMMessage
        message1.sender = createUser()

        uiMOC.processPendingChanges()

        // WHEN
        let lastMessage = conversation.lastMessageSent(by: selfUser)

        // THEN
        XCTAssertEqual(lastMessage, nil)
    }

    func testThatWeCanInsertAButtonActionMessage() {
        // GIVEN
        let conversation = ZMConversation.insertNewObject(in: uiMOC)
        conversation.remoteIdentifier = UUID()
        let buttonId = UUID().transportString()
        let messageId = UUID()

        // WHEN
        let message = try? conversation.appendButtonAction(havingId: buttonId, referenceMessageId: messageId)

        // THEN
        let expectedMessage = conversation.hiddenMessages.first
        XCTAssertEqual(message, expectedMessage)
        XCTAssertEqual(message?.underlyingMessage?.buttonAction.buttonID, buttonId)
        XCTAssertEqual(message?.underlyingMessage?.buttonAction.referenceMessageID, messageId.transportString())
    }
}
